tools/chpldef/Message.cpp

```diff
@@ -0,0 +1,294 @@
+#include <iostream>
+#include <fstream>
+
+namespace chpldef {
+
+bool Message::isIdValid(const JsonValue& id) {
+  return id.kind() == JsonValue::Number || id.kind() == JsonValue::String;
+}
+
+// TODO: Need to build an error message or record what went wrong. Create
+// an 'Invalid' subclass and use that to record any error that occurs past
+// the point of parsing the message ID.
+chpl::owned<Message> Message::request(Server* ctx, const JsonValue& json) {
+  auto objPtr = json.getAsObject();
+  if (!objPtr) {
+    ctx->verbose("Failed to unpack JSON object, found: %s\n",
+                 jsonTagStr(json));
+    return nullptr;
+  }
+
+  // Get the ID as a JSON value.
+  JsonValue id = nullptr;
+  if (auto idPtr = objPtr->get("id")) {
+    if (Message::isIdValid(*idPtr)) id = *idPtr;
+  }
+
+  // Failed to get the ID.
+  if (id.kind() == JsonValue::Null) {
+    ctx->verbose("Failed to get message ID\n");
+    return nullptr;
+  }
+
+  // Determine what method is to be called.
+  Message::Tag tag = Message::UNSET;
+  if (auto optMethod = objPtr->getString("method")) {
+    tag = Message::jsonRpcMethodNameToTag(optMethod->str());
+
+    // TODO: We can create an 'Invalid' request to represent this failure.
+    bool hasTag = tag != Message::UNSET && tag != Message::INVALID;
+    if (!hasTag) {
+      CHPLDEF_TODO();
+      return nullptr;
+    }
+  }
+
+  CHPL_ASSERT(tag != Message::RESPONSE);
+
+  // Before building a specific request, get the params.
+  const JsonValue* params = nullptr;
+  if (auto ptr = objPtr->get("params")) {
+    auto k = ptr->kind();
+    if (k != JsonValue::Array && k != JsonValue::Object) {
+      CHPLDEF_TODO();
+      return nullptr;
+    } else {
+      params = ptr;
+    }
+  }
+
+  switch (tag) {
+    #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+    case name__: return name__::create(std::move(id), *params);
+    #include "./message-macro-list.h"
+    #undef CHPLDEF_MESSAGE
+    default: break;
+  }
+
+  CHPLDEF_IMPOSSIBLE();
+
+  return nullptr;
+}
+
+// TODO: Is outbound format the same as for incoming requests?
+template <typename P, typename R>
+JsonValue Request<P, R>::pack() const {
+  return p.toJson();
+}
+
+template <typename T>
+static Response createSuccessfulResponse(const T* t) {
+  CHPL_ASSERT(t && t->isCompleted() && !t->hasError());
+  auto r = t->result();
+  CHPL_ASSERT(r);
+  auto data = r->toJson();
+  auto ret = Response::create(t->id(), std::move(data));
+  return ret;
+}
+
+opt<Response> Message::response(Server* ctx, const Message* msg) {
+  opt<Response> ret;
+
+  if (msg->tag() == Message::UNSET || msg->tag() == Message::INVALID ||
+      msg->tag() == Message::RESPONSE) {
+    return ret;
+  }
+
+  switch (msg->status()) {
+    case COMPLETED: {
+      CHPL_ASSERT(msg->error() == Message::OK);
+      switch (msg->tag()) {
+        #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+        case name__: return createSuccessfulResponse(msg->to##name__());
```

Comment:
I think I'd prefer all the `#define`s bodies to be indented relative to the rest of the code. Makes it easier to see what's part of the `#define`.

Reply:
T_T I'm the opposite, I find it weird to have the code in the define indented for no reason (relative to the non-expanded code. I can make the change though.

Reply:
Indented.